File: backend/internal/render/watermark_test.go
package render

import (
	"strings"
	"testing"

	"studio/internal/schema"
)

// Twins with frontend/src/watermark.test.ts.
func TestWatermarkLayoutGolden(t *testing.T) {
	g := watermarkLayout(&schema.Watermark{}, 200, 100, 1920, 1080)
	if g != (watermarkGeom{x: 1658, y: 934, w: 230, h: 114}) {
		t.Errorf("defaults = %+v, want {1658 934 230 114}", g)
	}
	tl := watermarkLayout(&schema.Watermark{Corner: "tl"}, 200, 100, 1920, 1080)
	if tl.x != 32 || tl.y != 32 {
		t.Errorf("tl = %+v, want margin 32", tl)
	}
	sq := watermarkLayout(&schema.Watermark{}, 0, 0, 1920, 1080)
	if sq.w != sq.h {
		t.Errorf("unknown dims should fall back to square, got %+v", sq)
	}
}

func TestWatermarkEntersTheGraph(t *testing.T) {
	doc := &schema.EditDoc{
		Canvas:    schema.Canvas{Width: 1280, Height: 720, FPS: 24},
		Assets:    []schema.Asset{{ID: "logo", Kind: "image", Width: 200, Height: 100}},
		Watermark: &schema.Watermark{AssetID: "logo", Opacity: 0.5},
		Tracks: []schema.Track{{ID: "v", Kind: schema.TrackVideo, Clips: []schema.Clip{{
			ID: "c1", AssetID: "a", Start: 0, In: 0, Out: 2,
			Transform: schema.Transform{Scale: 1, Opacity: 1},
		}}}},
	}
	args := compileArgs(t, doc)
	if !strings.Contains(args, "colorchannelmixer=aa=0.500[wmk]") {
		t.Errorf("watermark opacity missing from graph:\n%s", args)
	}
	if !strings.Contains(args, "[wmk]overlay=") {
		t.Errorf("watermark overlay missing from graph:\n%s", args)
	}

	doc.Watermark = nil
	if strings.Contains(compileArgs(t, doc), "wmk") {
		t.Error("no watermark configured, yet one entered the graph")
	}
}

File: frontend/src/watermark.test.ts
import { describe, it, expect } from "vitest";
import { watermarkLayout, watermarkOpacity } from "./watermark";

// Twins with render's TestWatermarkLayoutGolden.
describe("watermarkLayout golden", () => {
  it("defaults tuck a 2:1 logo into the bottom-right", () => {
    expect(watermarkLayout({ assetId: "a" }, 200, 100, 1920, 1080)).toEqual({ x: 1658, y: 934, w: 230, h: 114 });
  });

  it("corners are symmetric", () => {
    const tl = watermarkLayout({ assetId: "a", corner: "tl" }, 200, 100, 1920, 1080);
    const br = watermarkLayout({ assetId: "a", corner: "br" }, 200, 100, 1920, 1080);
    expect(tl.x).toBe(32);
    expect(tl.y).toBe(32);
    expect(br.x + br.w + 32).toBe(1920);
    expect(br.y + br.h + 32).toBe(1080);
  });

  it("unknown image dims fall back to square rather than nothing", () => {
    const g = watermarkLayout({ assetId: "a" }, 0, 0, 1920, 1080);
    expect(g.w).toBe(g.h);
  });

  it("opacity defaults and clamps", () => {
    expect(watermarkOpacity({ assetId: "a" })).toBe(0.6);
    expect(watermarkOpacity({ assetId: "a", opacity: 2 })).toBe(1);
  });
});
